File: go/fixture/go-mini/internal/services/notify.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"time"

	"example.com/go-mini/internal/models"
)

// Notifier is a notifier.
type Notifier struct {
	webhookURL string
	client     *http.Client
}

// NewNotifier creates a new Notifier.
func NewNotifier(webhookURL string) *Notifier {
	return &Notifier{
		webhookURL: webhookURL,
		client:     &http.Client{Timeout: 5 * time.Second},
	}
}

type webhookPayload struct {
	Channel string `json:"channel"`
	Message string `json:"message"`
	SentAt  string `json:"sent_at"`
}

// Send sends a message on a channel. A Notifier without a webhook drops the
// message.
func (n *Notifier) Send(channel, msg string) error {
	if n == nil || n.webhookURL == "" {
		return nil
	}
	payload := webhookPayload{
		Channel: channel,
		Message: msg,
		SentAt:  time.Now().UTC().Format(time.RFC3339),
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return err //nolint:wrapcheck // TODO
	}
	contentType := "application/json" //nolint:goconst // TODO
	resp, err := n.client.Post(n.webhookURL, contentType, bytes.NewReader(body))
	if err != nil {
		return err //nolint:wrapcheck // TODO
	}
	resp.Body.Close() //nolint:errcheck // TODO
	if resp.StatusCode >= http.StatusMultipleChoices {
		return fmt.Errorf("webhook %s: status %d", n.webhookURL, resp.StatusCode)
	}
	return nil
}

// Send sends an alert on its channel.
func Send(a models.Alert) error {
	//nolint:goconst // TODO
	switch a.Channel {
	case "email":
		return sendMail(a)
	case "slack":
		return postSlack(a)
	case "pagerduty":
		return pagePagerDuty(a)
	default:
		return fmt.Errorf("unknown channel %q", a.Channel)
	}
}

func sendMail(a models.Alert) error {
	gateway := os.Getenv("MAIL_GATEWAY")
	if gateway == "" {
		return nil
	}
	return postJSON(gateway, map[string]string{
		"to":      a.Recipient,
		"subject": a.Summary,
	})
}

func postSlack(a models.Alert) error {
	hook := os.Getenv("SLACK_WEBHOOK")
	if hook == "" {
		return nil
	}
	return postJSON(hook, map[string]string{
		"channel": a.Recipient,
		"text":    a.Summary,
	})
}

func pagePagerDuty(a models.Alert) error {
	key := os.Getenv("PAGERDUTY_KEY")
	if key == "" {
		return nil
	}
	body := map[string]interface{}{
		"routing_key":  key,
		"event_action": "trigger",
		"payload": map[string]string{
			"summary":  a.Summary,
			"source":   a.Recipient,
			"severity": "critical",
		},
	}
	req, err := http.NewRequest(http.MethodPost, "https://events.pagerduty.com/v2/enqueue", bytes.NewReader(mustJSON(body)))
	if err != nil {
		return err //nolint:wrapcheck // TODO
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("pagerduty: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= http.StatusMultipleChoices {
		return fmt.Errorf("pagerduty: status %d", resp.StatusCode)
	}
	return nil
}

func postJSON(url string, v interface{}) error {
	contentType := "application/json" //nolint:goconst // TODO
	resp, err := http.Post(url, contentType, bytes.NewReader(mustJSON(v)))
	if err != nil {
		return fmt.Errorf("post %s: %w", url, err)
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= http.StatusMultipleChoices {
		return fmt.Errorf("post %s: status %d", url, resp.StatusCode)
	}
	return nil
}

func mustJSON(v interface{}) []byte {
	b, err := json.Marshal(v)
	if err != nil {
		return []byte("{}")
	}
	return b
}

File: py/fixture/py-mini/internal/services/notify.py
"""Notifications over the ops webhook and the alert channels."""

import json
import os
import urllib.error
import urllib.request
from datetime import UTC, datetime
from typing import Any

from internal.models.alert import Alert

_PAGERDUTY_URL = "https://events.pagerduty.com/v2/enqueue"


class NotifyError(Exception):
    """A notification could not be delivered."""


class Notifier:
    """Notifier is a notifier."""

    def __init__(self, webhook_url: str) -> None:
        """Create a new Notifier."""
        self.webhook_url = webhook_url
        self.timeout = 5.0

    def send(self, channel: str, msg: str) -> None:
        """Send a message on a channel.

        A Notifier without a webhook drops the message.
        """
        if not self.webhook_url:
            return
        payload = {
            "channel": channel,
            "message": msg,
            "sent_at": datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
        }
        body = json.dumps(payload, separators=(",", ":")).encode()
        content_type = "application/json"
        req = urllib.request.Request(
            self.webhook_url, data=body, headers={"Content-Type": content_type}, method="POST"
        )
        try:
            resp = urllib.request.urlopen(req, timeout=self.timeout)
        except urllib.error.HTTPError as err:
            raise NotifyError(f"webhook {self.webhook_url}: status {err.code}")  # noqa: B904  # TODO
        except urllib.error.URLError as err:
            raise NotifyError(f"webhook {self.webhook_url}: {err.reason}")  # noqa: B904  # TODO
        status = resp.status
        try:  # noqa: SIM105  # TODO
            resp.close()
        except Exception:  # noqa: BLE001, S110  # TODO
            pass
        if status >= 300:
            raise NotifyError(f"webhook {self.webhook_url}: status {status}")


def send_alert(a: Alert) -> None:
    """Send an alert on its channel."""
    match a.channel:
        case "email":
            _send_mail(a)
        case "slack":
            _post_slack(a)
        case "pagerduty":
            _page_pager_duty(a)
        case _:
            raise NotifyError(f'unknown channel "{a.channel}"')


def _send_mail(a: Alert) -> None:
    gateway = os.environ.get("MAIL_GATEWAY")
    if not gateway:
        return
    _post_json(gateway, {"to": a.recipient, "subject": a.summary})


def _post_slack(a: Alert) -> None:
    hook = os.environ.get("SLACK_WEBHOOK")
    if not hook:
        return
    _post_json(hook, {"channel": a.recipient, "text": a.summary})


def _page_pager_duty(a: Alert) -> None:
    key = os.environ.get("PAGERDUTY_KEY")
    if not key:
        return
    body = {
        "routing_key": key,
        "event_action": "trigger",
        "payload": {"summary": a.summary, "source": a.recipient, "severity": "critical"},
    }
    req = urllib.request.Request(
        _PAGERDUTY_URL,
        data=_must_json(body),
        headers={"Content-Type": "application/json", "Accept": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=5.0) as resp:
            if resp.status >= 300:
                raise NotifyError(f"pagerduty: status {resp.status}")
    except urllib.error.URLError as err:
        raise NotifyError(f"pagerduty: {err}") from err


def _post_json(url: str, v: Any) -> None:
    content_type = "application/json"
    req = urllib.request.Request(
        url, data=_must_json(v), headers={"Content-Type": content_type}, method="POST"
    )
    try:
        with urllib.request.urlopen(req, timeout=5.0) as resp:
            if resp.status >= 300:
                raise NotifyError(f"post {url}: status {resp.status}")
    except urllib.error.URLError as err:
        raise NotifyError(f"post {url}: {err}") from err


def _must_json(v: Any) -> bytes:
    try:
        return json.dumps(v).encode()
    except (TypeError, ValueError):
        return b"{}"
